ENH: Add intuitive volume presets

rbumm opened the pull request:
Add functionality to display volume presets from the slice view right-click menu.  
Do not display the preset menu in 3D view.   
Let it only display if a scalarvolume is displayed in slice view.   
Make the presets checkable and setcheck(true) them upon selection  
Make this work even if several scalar volumes are loaded.  
Set the presets by a function call to  [Modules/Loadable/Volumes/Logic/vtkSlicerVolumesLogic.cxx](https://github.com/Slicer/Slicer/compare/master...rbumm:master#diff-e66332c76fe962dd310d54e008ed379b062f3fec8a74f0969d4706d96b362ad4) 

![](https://user-images.githubusercontent.com/18140094/139442405-7fab1660-7414-4cb1-9aa8-15c237f53f51.png)

![](https://user-images.githubusercontent.com/18140094/139446279-48726161-2584-45cd-b0ed-36b6c482236c.png)

To do:

find a way to define a default window/level preset that the user can define by special action and that be used as default upon every start of Slicer.  

Limitations:

*   qSlicerScalarVolumeDisplayWidget has still its hard coded setPreset() function, pls advise what to do here
*   setting the colornode seems to have different results if done from the right click menu compared to qSlicerScalarVolumeDisplayWidget.

rbumm commented on Modules/Loadable/SubjectHierarchy/Widgets/qSlicerSubjectHierarchyViewContextMenuPlugin.h:
```diff
@@ -50,6 +50,15 @@ class Q_SLICER_MODULE_SUBJECTHIERARCHY_WIDGETS_EXPORT qSlicerSubjectHierarchyVie
+       ID_CTBrain,
+       ID_CTLung,
+       ID_DTI,
+  };
```
Great workflow and got several aspects realized,  but unfortunately, I am unable to call 

volumeModulePointer = d->ModuleManager->module("Volumes");

from  qSlicerSubjectHierarchyViewContextMenuPluginPrivate::init() because the module "Volume" seems not yet to have loaded at the time when init() is executed. Any ideas or can I somehow directly reference the module and the connected logic()?

pieper replied:
It is possible for modules to express dependencies, but it's not a common requirement and probably better to avoid it.
https://github.com/Slicer/Slicer/blob/0094e9a35bd266cc8b0e677858dabce797c9928f/Base/QTCore/qSlicerAbstractCoreModule.h#L154-L160

For this reason and others I'd still suggest moving the preset information into the display node instead of having the subject hierarchy plugin access the volume module (in general we'd want all mrml operations to work even if modules and especially module guis are not instantiated.

rbumm replied:
> modules to express dependencies

Thanks, bút I was not able to define dependencies in qSlicerSubjectHierarchyViewContextMenuPlugin, probably because it is not a loadable module.

lassoan replied:
I just realized that you modified the common base class qSlicerSubjectHierarchyViewContextMenuPlugin.cxx. That is only for adding generic view menu actions. Volumes related actions should be added to qSlicerSubjectHierarchyVolumesPlugin.cxx in viewContextMenuActions() method, similarly how it is done for markups in qSlicerSubjectHierarchyMarkupsPlugin::viewContextMenuActions().

lassoan commented on Modules/Loadable/Volumes/Logic/vtkSlicerVolumesLogic.cxx:
```diff
@@ -1545,3 +1553,128 @@ ::ResampleVolumeToReferenceVolume(vtkMRMLVolumeNode* inputVolumeNode,
+}
+
+// --------------------------------------------------------------------------
+void vtkSlicerVolumesLogic::setWindowLevelPreset(vtkMRMLScalarVolumeNode* volumeNode, const QString& presetName)
```
Yes, the display node definitely should be able to store some volume-specific presets, for example the preset(s) imported from the DICOM file.

It would be tempting to store application-wide presets in the display node, too, but that would not be consistent with how we store almost all other presets/constants in module logics (color tables, volume rendering presets, terminologies, and units). Slice view orientation preset management is an exception (those are stored in the scene), and it is not a very clean or robust design. For example, if we want to change orientation presets then we need to iterate through all existing slice nodes and modify all of them, which is somewhat complicated and it also removes all customizations that other modules may have applied. You would probably also need to update the window/level presets in all volume display nodes when you load a scene, which would not be nice.

View context menu is managed by Subject Hierarchy plugins, which can access any module logic class. See for example [here](https://github.com/Slicer/Slicer/blob/7eed7c54ea41b2a5e3c4b73c8e5d84391876e71f/Modules/Loadable/Markups/SubjectHierarchyPlugins/qSlicerSubjectHierarchyMarkupsPlugin.cxx#L291), so it should be no problem to maintain a shared list of window/level presets there.

pieper replied:
I wouldn't suggest saving the preset values in the as instance variables in display nodes (that is, they shouldn't be saved in the scene), only that any hard-coded values for certain displayable types should be in the display nodes and that any helper classes to manipulated them should be in the logics.  Keeping the options in the display node code also makes sense since they can then easily be augmented or overwritten by subclasses.

lassoan replied:
How do you envision it would work? Would the data reader determine what kind of image is loaded (CT, US, MRI, PET,...custom image types) and initialize the presets in the display node?

If we don't save the presets in the scene then how would the presets would be set in display nodes that are loaded from a scene?

pieper replied:
What I mean is that currently these hard-coded preset defaults are [stored in the widget code](https://github.com/Slicer/Slicer/blob/550f0334a3538609e4748ec18ec77bfaddda7950/Modules/Loadable/Volumes/Widgets/qSlicerScalarVolumeDisplayWidget.cxx#L406-L476_), which is clearly the wrong place.  I think it would be cleaner to have this information provided by the display node (as class-static data).  This would be in addition to the existing window level preset list, which is in an instance variable and is saved with the scene.  Then if we did make, for example, a `vtkMRMLPETVolumeNode` with a `vtkMRMLPETVolumeDisplayNode` the display node could provide only the presets that make sense for the specialized subclass.

lassoan replied:
How would it work for CT? Would you create a new vtkMRMLCTVolumeDisplayNode and use that instead of vtkMRMLScalarVolumeDisplayNode when you detect that it is a CT?

An image modality detection mechanism would help in recommending volume rendering presets and color node, too. This logic is currently implemented in the volume rendering subject hierarchy plugin, but it would make sense to implement it at lower level (MRML and/or module logic).

In the future, modality could be used to recommend hanging protocol and analysis modules.

pieper replied:
We have this pattern already for DiffusionWeighted and DiffusionTensor volumes, where they override certain behaviors in the slice views and UI.  But it's incomplete since some of the Volumes UI is hard coded.

So yes, I can imagine having a specialized volumes for modalities that expose custom rendering options.  We already do some of this when reading from dicom.  When we read from nrrd or other formats this information is lost, but we could offer the user options to specify it when known (like we do when reading a nifti as a segmentation for example).

rbumm commented:
I pushed a commit today as a follow-up to yesterday's interesting discussion and addressed some of the requests:   
Presets are now being read from vtkSlicerVolumesLogic  (still hardcoded in a std::vector)  
Moved the menu actions to qSlicerSubjectHierarchyVolumesPlugin upon @lassoan remark. 

All works well and pressing the 'Volumes' preset buttons now check the right click menu.  Finally, please instruct me how to realize the JSON preset storage idea (where to put the preset file, how would I add it to the project, would you think it needs to be a read/write or read-only). Thanks !

lassoan commented:
Such configuration files are typically stored in the Slicer install tree in the module share directory (`share\Slicer-4.13\`). See how VolumeRendering and Terminologies modules discover and load their presets.

pieper commented:
Great progress!

(haha, Andras beat me on the suggestion below by a minute)
For the json file, it could go in the `Modules/Loadable/Volumes/Resources` directory so it can be bundled with the application to a predictable spot.  You can see examples of that in `Markups` and `Terminologies` modules.

rbumm commented:
I pushed a commit today which realizes reading the presets from a JSON file.   
In addition, the 'Volumes' preset buttons now "check" and action in the qSlicerScalarVolumeDisplayWidget ´s MRMLWindowLevelWidget no "unchecks" the menu actions.

lassoan commented:
@rbumm I've reviewed the code and added a few more features and cleaned up a few things (see details in the commit). Please give it a test, but I think this should be good to go.

There are many possibilities for further improvements, such as generating thumbnails, or adding window/level preset submenu for each volume in the mouse mode selection toolbar, etc. but these could be implemented later.

rbumm commented:
> Please give it a test

@lassoan Will test tomorrow. I used the enhancement regularly for about a week now and I am quite happy with the function. Would you think we should implement saving the last window/level setting as soon as the user selects a preset from the menu or presses one of the preset button so that a new slicer instance would load scalar volumes with the last saved setting?

lassoan commented:
Blindly applying a fixed preset to all loaded images would only work for very specific usage patterns of Slicer, so it could not be a Slicer core feature.

One option would be to make window/level (and maybe also view layout, etc. customizations) to a lung CT analysis module. Once you open that module and an image is selected, it would set the display settings that are relevant for that analysis.

This feature could be generalized by adding a "protocol" selector to Slicer core, which would allow the user to choose how the images should be used: for example, do a lung CT density analysis, a brain tumor volumetry, DCE-MRI prostate analysis, etc. and Slicer would set up views, display properties, and start a module (or offer a list of relevant modules). The protocol selection could be automatic, or semi-automatic (protocols could be suggested based on what imaging modalities are found the study, series descriptions, etc.).

pieper reviewed (approved):
I didn't try it, but it looks good to me.

pieper commented:
@rbumm do you typically load from DICOM?  If so the W/L should be coming from the headers.  Is that not working for you?

rbumm commented:
> @rbumm do you typically load from DICOM? If so the W/L should be coming from the headers. Is that not working for you?

Yes, I import from DICOM but I usually get a slightly different windows/level combination which is less bright right and has more contrast than what I am used to when pressing the Lung Prest button. The series from @pzaffino comes as nrrd and is very bright with low contrast initially and also needs the CT-Lung preset to be applied.

pieper commented:
Yes, makes sense.  Nrrd files don't have w/l presets.  At some point it could work to make a network that recognizes image contents and suggests good w/l settings automatically - could be a fun project for someone getting going with ML.

lassoan commented:
I've added window/level presets from the display node (e.g., those that are read from DICOM) and force-pushed the changes to this branch:

![image](https://user-images.githubusercontent.com/307929/141691213-872c2ec0-cbba-4202-ad4b-cadac13bf2fa.png)

rbumm commented:
Looks good to me, but need to clone my fork again to the dev computer to get this running, and is now in the presumably 6 h build. 

I did not succeed in simply pulling the changed sources, they compiled, but I gut linker errors thereafter.

lassoan commented:
The circleCI build test succeeded, so most likely there was some error in your build tree.

I'll merge it then and you can test it out in the Slicer Preview Release tomorrow.

lassoan commented:
@rbumm This feature should be available in Slicer Preview Releases downloaded on Monday or later. It would be great if you could test it and if it works well then write a short post on the Slicer forum, in "Support" category, with "feature" tag added, with a short summary of the feature (the [commit comment ](https://github.com/Slicer/Slicer/commit/c72d890515ccca70e4173dfcedd370e197c44ae9) may help) and a screenshot.

rbumm commented:
Thank you @lassoan , I was able to compile after the new build, and see the new menu entry. 

Unfortunately, at least here when pressing the preset buttons in "Volumes" no longer work, in Slicer's secondary window I see: 

```
vtkSlicerVolumesLogic::ApplyVolumeDisplayPreset failed: Could not find preset ID CTBonePresetToolButton
```

rbumm commented:
Update: The "Volume" preset buttons do not work in today's Slicer preview, all other changes are there.

rbumm commented:
@lassoan got it working by replacing 

```
// --------------------------------------------------------------------------
void qSlicerScalarVolumeDisplayWidget::onPresetButtonClicked()
{
  QToolButton* preset = qobject_cast<QToolButton*>(this->sender());
  this->setPreset(preset->objectName());
}
```

with

```
// --------------------------------------------------------------------------
void qSlicerScalarVolumeDisplayWidget::onPresetButtonClicked()
{
  QToolButton* preset = qobject_cast<QToolButton*>(this->sender());
  this->setPreset(preset->accessibleName());
}
```

Would be glad if you could implement this change, thank you\`

jamesobutler commented:
Note there are 3 new failed tests today following the commit merged from this PR
https://slicer.cdash.org/viewTest.php?onlyfailed&buildid=2476970

qSlicerVolumesModuleGenericTest
qSlicerVolumesModuleWidgetGenericTest
qSlicerVolumesModuleWidgetTest1

lassoan commented:
Thanks, I'll fix them.

lassoan commented:
All done - https://github.com/Slicer/Slicer/commit/409c7f035aa9015863cffc827ed37661322a550d

Presets in Volumes module are now read from the VolumeDisplayPresets.json file. Therefore, presets in both the Volumes module and slice view context menu are the same and presets can be added/reordered/removed/modified by editing the json file.

In addition to this, also fixed the failing tests, and added icons to the slice view context menu for consistency (and slightly more interesting look).

![image](https://user-images.githubusercontent.com/307929/141857853-d0491e5c-253c-4b06-a172-eb588dc498bd.png)

rbumm commented:
Wow. Very nice.
